src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java

```diff
@@ -0,0 +1,212 @@
+     */
+    public Matrix<N3, N1> calculateStandardDeviations() {
+        if (isWithinBestTagRangeForSolvePNP())
+            return calculateSolvePNPStandardDeviations();
+        return calculateAssumedHeadingStandardDeviations();
+    }
+
+    /**
+     * If the robot is close enough to the tag, it gets the estimated solve PNP pose.
+     * If it's too far, it assumes the robot's heading and calculates its position from there.
+     * Assuming the robot's heading is more robust, but won't fix current wrong heading.
+     * To fix this, we use solve PNP to reset the robot's heading when we are close enough for an accurate result.
+     *
+     * @return the robot's pose
+     */
+    private Pose2d calculateBestRobotPose() {
+        final Rotation2d gyroHeadingAtTimestamp = PoseEstimator6328.getInstance().samplePose(inputs.latestResultTimestampSeconds).getRotation();
+        return calculateAssumedRobotHeadingPose(isWithinBestTagRangeForSolvePNP() ? inputs.solvePNPHeading : gyroHeadingAtTimestamp);
+    }
+
+    /**
+     * Calculates the robot's pose by assuming its heading, the apriltag's position, and the camera's position on the robot.
+     * This method of finding the robot's pose is more robust but relies on knowing the robot's heading beforehand.
+     *
+     * @return the robot's pose
+     */
+    private Pose2d calculateAssumedRobotHeadingPose(Rotation2d robotHeading) {
+        if (inputs.visibleTagIDs.length == 0)
+            return null;
+
+        final Translation2d robotFieldRelativeTranslation = getRobotFieldRelativeTranslation(robotHeading);
+        return new Pose2d(robotFieldRelativeTranslation, robotHeading);
+    }
+
+    private Transform2d toTransform2d(Transform3d transform3d) {
+        return new Transform2d(
+                transform3d.getTranslation().toTranslation2d(),
+                transform3d.getRotation().toRotation2d()
+        );
+    }
+
+    private Translation2d getRobotFieldRelativeTranslation(Rotation2d robotHeading) {
+        if (!inputs.hasResult)
+            return new Translation2d();
+
+        Pose2d fieldRelativeRobotPose = PhotonUtils.estimateFieldToRobot(
+                robotCenterToCamera.getZ(),
+                AprilTagCameraConstants.TAG_ID_TO_POSE.get(inputs.visibleTagIDs[0]).getZ(),
+                -robotCenterToCamera.getRotation().getY(),
+                -inputs.bestTargetRelativePitchRadians,
+                Rotation2d.fromRadians(inputs.bestTargetRelativeYawRadians),
+                robotHeading,
+                AprilTagCameraConstants.TAG_ID_TO_POSE.get(inputs.visibleTagIDs[0]).toPose2d(),
+                toTransform2d(robotCenterToCamera.inverse())
+        );
+        return fieldRelativeRobotPose.getTranslation();
+    }
+
+    private boolean isNewTimestamp() {
+        if (lastUpdatedTimestamp == getLatestResultTimestampSeconds())
+            return false;
+
+        lastUpdatedTimestamp = getLatestResultTimestampSeconds();
+        return true;
+    }
```

Comment:
you can just first update `lastUpdatedTimestamp` and then `return lastUpdatedTimeStamp == getLatestResultTimestampSeconds();`
Also, maybe add a tolerance

Reply:
We can't because it checks the current timestamp to the last timestamp, and if it's different, it overwrites it in the same variable. Wdym tolerance? It's checking the time